---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# stmut

<!-- badges: start -->
<!-- badges: end -->

The goal of stmut is providing a series of functions to visualize which genes have somatic mutations in which spots when working on sptial transcriptomics data.



## Installation

You can install the development version of stmut from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("limin321/stmut")
```
## Notes before running the package.
This package has a series of 4 functions: sptBClstRds, sptMutCt, nonZeRdCts, spotSummary. <br />
You have to run each function sequentially to obtain the ideal output, or You have your input of each function strictly match the input format of our examples.



Preparations -- 5 files from spaceranger output: <br />
filtered_feature_bc.csv <br />
Graph-Based.csv <br />
possorted_genome_bam.bam <br />
spatial/tissue_positions_list.csv <br />
raw_feature_bc_matrix/barcodes.tsv.gz <br />



spotIndex: one spatial barcode one .txt file, thie file name is the spot name you choose. <br />

spotBam:
https://github.com/10XGenomics/subset-bam <br />
subset-bam_linux --bam possorted_genome_bam.bam --cell-barcodes spot000.txt --out-bam spot000.bam
samtools index spot000.bam

Count each spot ref and mut reads using Mpileup_RNA.pl (samtools mpileup): <br />
perl Mpileup_RNA.pl somaticSNPsMML.txt spot000/MpileupOutput_RNA.txt






## Example

This is a basic example of how to run sptBClstRds function:

```{r example}
 library(stmut)
## basic example code
files <- list.files(path = "./inst/extdata/spotIndex", pattern = ".txt", full.names = TRUE, recursive = FALSE)
data1 <- read.csv("./inst/extdata/filtered_feature_bc.csv", header = TRUE)
data2 <- read.csv("./inst/extdata/Graph-Based.csv", header = TRUE)
path <- "./inst/extdata/tissue_positions_list.csv"

df <- sptBClstRds(files=files,data1=data1, data2=data2, path=path)

head(df[[1]])
head(df[[2]])
```


An example on how to run sptMutCt()
```{r example2}
library(stmut)
index <- read.csv("./inst/extdata/spotBC.csv", header = TRUE)
files2 <- list.files(path = "./inst/extdata/mpileup",pattern = "MpileupOutput_RNA.txt", full.names = TRUE, recursive = TRUE, include.dirs = TRUE)
df1 <- sptMutCt(index = index, files = files2)

head(df1[[1]])
head(df1[[2]])


```

An example to run nonZeRdCts
```{r example3}
library(stmut)
d1 <- read.csv("./inst/extdata/spotBClster.csv", header = TRUE)
d2 <- read.csv("./inst/extdata/spotRdCtFinal.csv", header = TRUE)

dim(d1)
dim(d2)

df3 <- nonZeRdCts(df1 = d1, df2 = d2)
head(df3[[1]])
head(df3[[2]])

```
An example to run spotSummary
```{r example4}
library(stmut)
df <- read.csv("./inst/extdata/NonZeroRdSpotIndex.csv", header = TRUE)
path1 <-"./inst/extdata/mpileup/"

df4 <- spotSummary(df = df, path1 = path1)

head(df4)

```
Column meaning of df4: <br />
column 1: Spatial barcode <br />
column 2: spot name <br />
column 3: cluster in Loupe.Loupe file <br />
column 4: Total reads each spot has <br />
column 5: row position in position_lists_bc.csv <br />
column 6: col position in position_lists_bc.csv <br />
column 7: Reference read counts <br />
column 8: Mutant read counts <br />
column 9: Sum of Ref read count and Mut read count (sum of col 7 and col 8) <br />
column 10: The number ( < 10) in this column indicates how many mutant reads in a spot. For example, G9 means there are 9 mutant reads in that spotl G10 means there are more than 10 mutant reads in the spot. <br />
column 11: proportion of mutant to Total reads (col8/col4) <br />
column 12: proportion of mutant to sum of ref and mut reads (col8/col9) <br />
column 13: A binary column, when the spot has > 1 mutant read, it is 1, otherwise, it is 0. <br />
column 14: A score calculated based on the ref and mutant read counts. The higher the score is, the more likely it is a tumor spot. The table is ordered by this column. <br />
column 15: Details on which genes containing what mutations in which spots. <br />



```{r, include=FALSE}
# You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. You could also use GitHub Actions to re-render `README.Rmd` every time you push. An example workflow can be found here: <https://github.com/r-lib/actions/tree/v1/examples>.
```

